cloudant-http/src/main/java/com/cloudant/http/internal/interceptors/CookieInterceptorBase.java

```diff
@@ -0,0 +1,294 @@
+        return context;
+
+    }
+
+    // helper for requestCookie
+    interface OnExecuteCallable {
+        boolean call(HttpConnection connection) throws IOException;
+    }
+
+    protected boolean requestCookie(HttpConnectionInterceptorContext context,
+                                    URL url,
+                                    byte[] payload,
+                                    String mimeType,
+                                    String accept,
+                                    OnExecuteCallable onResponseOk) {
+
+        try {
+            HttpConnection conn = Http.POST(url, mimeType);
+            conn.requestProperties.put("accept", accept);
+            conn.setRequestBody(payload);
+
+            //when we request the session we need all interceptors except this one
+
+            conn.requestInterceptors.addAll(context.connection.requestInterceptors);
+            conn.requestInterceptors.remove(this);
+            conn.responseInterceptors.addAll(context.connection.responseInterceptors);
+            conn.responseInterceptors.remove(this);
+
+            HttpConnection connection = conn.execute();
+            int responseCode = connection.getConnection().getResponseCode();
+
+            if (responseCode / 100 == 2) {
+                return onResponseOk.call(connection);
+            } else {
+                InputStream errorStream = connection.getConnection().getErrorStream();
+                try {
+                    if (errorStream != null) {
+                        // Consume the error stream to avoid leaking connections
+                        String error = IOUtils.toString(errorStream, "UTF-8");
+                        // Log the error stream content
+                        logger.fine(error);
+                    }
+                } finally {
+                    IOUtils.closeQuietly(errorStream);
+                }
+                // TODO these messages should be customised so we can tell if bearer or session token failed
+                if (responseCode == 401) {
+                    logger.severe("Credentials are incorrect, cookie authentication will not be" +
+                            " attempted again by this interceptor object");
+                } else {
+                    // catch any other response code
+                    logger.log(Level.SEVERE,
+                            "Failed to get cookie from server, response code {0}, " +
+                                    "cookie authentication will not be attempted again",
+                            responseCode);
+                }
+            }
+        }  catch (IOException e) {
+            logger.log(Level.SEVERE, "Failed to read cookie response", e);
+        }
+        return false;
+    }
+
+    protected boolean requestCookie(HttpConnectionInterceptorContext context) {
+        return requestCookie(context,
+                sessionURL,
+                sessionRequestBody,
+                "application/x-www-form-urlencoded",
```

Comment:
I think we should default to JSON here if we can as the encoding is likely more reliable. I am expecting `_iam_session` to accept JSON **if** the `oidc/token` can as well then it leaves the normal `_session` as the odd one out that can do an override.